frontend/lib/resolution/Resolver.cpp

```diff
@@ -2943,6 +2943,130 @@ bool Resolver::enter(const ReduceIntent* reduce) {
+
+  // Couldn't resolve the call; is opName a valid reduction?
+  if (opType.isUnknown()) {
+    CHPL_REPORT(context, ReductionInvalidName, reduce, opName, iterType);
+    return nullptr;
+  }
+
+  // We found some type; is it a subclass of ReduceScanOp?
+  auto baseClass = BasicClassType::getReduceScanOpType(context);
+  auto actualClass = opType.type()->toClassType();
+  bool converts, instantiates;
+  if (opType.kind() != QualifiedType::TYPE ||
+      !actualClass ||
+      !actualClass->basicClassType()->isSubtypeOf(baseClass, converts, instantiates)) {
+    CHPL_REPORT(context, ReductionNotReduceScanOp, reduce, opType);
+  }
+
+  return actualClass;
+}
+
+static const ClassType* determineReduceScanOp(Resolver& resolver,
+                                              const uast::Reduce* reduce,
+                                              const QualifiedType& iterType) {
+  if (auto ident = reduce->op()->toIdentifier()) {
+    auto toLookUp = ident->name();
+    auto opName = identifierReduceScanOpName(resolver.context, ident->name());
+    if (!opName.isEmpty()) {
+      // The identifier does not itself name a ReduceScanOp class, but is
+      // associated with such a class. Use the associated class' name instead
+      // of the identifier itself.
+      toLookUp = opName;
+    }
+    auto scanOp = constructReduceScanOpClass(resolver, reduce, toLookUp, iterType);
+    if (scanOp != nullptr) {
+      // Since we found a ReduceScanOp, set the refersToId of the identifier.
+      resolver.validateAndSetToId(resolver.byPostorder.byAst(ident),
+                                  ident, scanOp->basicClassType()->id());
+    }
+    // No further processing is needed; we found the operation.
+    return scanOp;
+  }
+
+  // In the future, node could be a value or something else, and this function
+  // would return the corresponding ClassType.
+
+  return nullptr;
+}
+
+static QualifiedType getReduceScanOpResultType(Resolver& resolver,
+                                               const uast::Reduce* reduce,
+                                               const ClassType* opClass) {
+  auto context = resolver.context;
+
+  auto borrowedDecorator =
+    ClassTypeDecorator(ClassTypeDecorator::BORROWED_NONNIL);
+  auto borrowedClass = opClass->withDecorator(context,
+                                              std::move(borrowedDecorator));
+  auto thisActual = QualifiedType(Qualifier::CONST_IN, borrowedClass);
+
+  std::vector<CallInfoActual> typeActuals;
+  typeActuals.push_back(CallInfoActual(thisActual, USTR("this")));
+  auto ci = CallInfo (/* name */ USTR("generate"),
+                      /* calledType */ thisActual,
+                      /* isMethodCall */ true,
+                      /* hasQuestionArg */ false,
+                      /* isParenless */ false,
+                      typeActuals);
+  const Scope* scope = scopeForId(context, reduce->id());
+  auto c = resolveGeneratedCall(context, reduce, ci, scope, resolver.poiScope);
```

Comment:
Perhaps this (or other reduce/scan details) should be saved in the AssociatedActions

Reply:
What would that look like? I see `handleResolvedAssociatedCall`, and I assume that I should be doing something similar. However, a couple of questions:
* Is there any more information we need to store here other than the result of resolving the generated call?
* Which expression should have the associated action, the whole `reduce` expression or just the `op` sub-expression?

Reply:
Yes, you'd save the call resolution result.

I think those are up for debate. I also think it's very reasonable to add a TODO comment to figure that out later.

One of the main reasons to save the resolved call is so that later analysis working on uAST + resolution results can correctly compute a call graph without having to have custom logic for reductions.

It is also arguably useful for making the uAST + resolution results -> old AST translation easier.